source/components/InfiniteScroll.js

```diff
@@ -0,0 +1,142 @@
+// @flow
+import React, { Component } from 'react';
+import type { ComponentType, Element } from 'react';
+
+// $FlowFixMe
+import createRef from 'create-react-ref/lib/createRef';
+
+// internal components
+import { withTheme } from './HOC/withTheme';
+
+// utilities
+import { composeTheme, addThemeId } from '../utils/themes';
+
+// constants
+import { IDENTIFIERS } from '../themes/API';
+
+type Props = {
+  className: string,
+  context: {
+    theme: Object,
+    ROOT_THEME_API: Object
+  },
+  fetchData: Function,
+  renderItems: Function,
+  skin: ComponentType<any>,
+  theme: Object, // will take precedence over theme in context if passed
+  themeId: string,
+  themeOverrides: Object,
+  threshold: number
+};
+
+type State = {
+  composedTheme: Object,
+  data: Object | Array<{}>,
+  error: boolean | string | Element<*>,
+  hasMoreData: boolean,
+  isLoading: boolean
+};
+
+class InfiniteScrollBase extends Component<Props, State> {
+  // declare ref types
+  scrollContainer: ?Element<any>;
+
+  // define static properties
+  static displayName = 'InfiniteScroll';
+  static defaultProps = {
+    fetchData() {},
+    theme: null,
+    themeId: IDENTIFIERS.INFINITE_SCROLL,
+    themeOverrides: {},
+    threshold: 250
+  };
+
+  constructor(props: Props) {
+    super(props);
+    const { context, themeId, theme, themeOverrides } = props;
+
+    // refs
+    this.scrollContainer = createRef();
+
+    this.state = {
+      composedTheme: composeTheme(
+        addThemeId(theme || context.theme, themeId),
+        addThemeId(themeOverrides, themeId),
+        context.ROOT_THEME_API
+      ),
+      data: [],
+      error: false,
+      isLoading: false,
+      hasMoreData: true
+    };
+  }
+
+  componentWillMount() {
+    this.handleFetchData();
+  }
+
+  componentDidMount() {
+    const { scrollContainer } = this;
+    scrollContainer.current.addEventListener('scroll', this.handleScroll);
+  }
+
+  _isFunction = (renderProp: ?Function) => (renderProp && typeof renderProp === 'function');
```

Comment:
@DominikGuzei 
I pushed the requested changes up. I removed the `setError` method as well because it doesn't get used. The user can set errors in the `fetchData` function passed a prop.